File: scripts/import_themealdb.py
#!/usr/bin/env python3
"""Import vegetarian + vegan recipes from TheMealDB (free, open recipe API)
into the app's recipe format.

TheMealDB: https://www.themealdb.com/api.php  (test key "1" used here)
Output: src/data/recipes/themealdb.json  (gitignored, merged into the catalog).

Stdlib only — no pip install needed.

Usage:
    python3 scripts/import_themealdb.py
"""

from __future__ import annotations

import json
import re
import time
import urllib.request
from pathlib import Path

API = "https://www.themealdb.com/api/json/v1/1"
OUT = Path(__file__).resolve().parent.parent / "src" / "data" / "recipes" / "themealdb.json"

# TheMealDB categories that fit a vegetarian app and their diet tag.
CATEGORIES = {"Vegetarian": "vegetarian", "Vegan": "vegan"}


def get(url: str) -> dict:
    req = urllib.request.Request(url, headers={"User-Agent": "gymapp-importer"})
    with urllib.request.urlopen(req, timeout=20) as r:
        return json.loads(r.read().decode("utf-8"))


def slug(s: str) -> str:
    return "tmdb-" + re.sub(r"[^a-z0-9]+", "-", s.lower()).strip("-")


def meal_category(strCategory: str) -> str:
    c = (strCategory or "").lower()
    if "breakfast" in c:
        return "breakfast"
    if "dessert" in c or "starter" in c or "side" in c:
        return "snack"
    return "dinner"


def ingredients(meal: dict) -> list[str]:
    out = []
    for i in range(1, 21):
        ing = (meal.get(f"strIngredient{i}") or "").strip()
        mea = (meal.get(f"strMeasure{i}") or "").strip()
        if ing:
            out.append((f"{mea} {ing}").strip())
    return out


def steps(instructions: str) -> list[str]:
    text = (instructions or "").replace("\r", "\n")
    parts = [p.strip() for p in re.split(r"\n+", text) if p.strip()]
    if len(parts) <= 1:  # one block — split into sentences
        parts = [s.strip() for s in re.split(r"(?<=[.!?])\s+", text) if s.strip()]
    return parts[:20]


def to_recipe(meal: dict, diet: str) -> dict:
    title = meal.get("strMeal", "Recipe")
    tags = [t.strip() for t in (meal.get("strTags") or "").split(",") if t.strip()]
    diet_tags = ["vegetarian"] if diet == "vegan" else [diet]
    if diet == "vegan":
        diet_tags.insert(0, "vegan")
    return {
        "id": slug(title),
        "title": title,
        "category": meal_category(meal.get("strCategory", "")),
        "minutes": 30,  # TheMealDB has no time; sensible default
        # TheMealDB has no macros — left at 0, shown as "—" in the UI.
        "kcal": 0, "protein": 0, "carbs": 0, "fat": 0,
        "ingredients": ingredients(meal),
        "steps": steps(meal.get("strInstructions", "")),
        "thumbnail": meal.get("strMealThumb") or None,
        "tags": list(dict.fromkeys([*tags, *diet_tags, "imported"])),
        "diet": diet_tags,
        "source": meal.get("strSource") or f"https://www.themealdb.com/meal/{meal.get('idMeal')}",
    }


def main() -> None:
    recipes: dict[str, dict] = {}
    for cat, diet in CATEGORIES.items():
        listing = get(f"{API}/filter.php?c={cat}").get("meals") or []
        print(f"{cat}: {len(listing)} meals")
        for stub in listing:
            mid = stub["idMeal"]
            detail = get(f"{API}/lookup.php?i={mid}").get("meals")
            if not detail:
                continue
            r = to_recipe(detail[0], diet)
            recipes.setdefault(r["id"], r)  # dedupe by id
            time.sleep(0.15)  # be polite to the free API

    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps({"recipes": list(recipes.values())}, indent=2))
    print(f"Wrote {len(recipes)} recipes -> {OUT.relative_to(OUT.parents[3])}")


if __name__ == "__main__":
    main()

File: scripts/import-themealdb.mjs
// Import TheMealDB (free recipe API) into our self-hosted pipeline.
// Fetches all meals (a–z), downloads each thumbnail locally (independence-gate
// safe), maps to our recipe schema with source:'themealdb', and writes
// downloaded_pages/themealdb_recipes.json for build-catalog to merge.
// Attribution required (TheMealDB) — commercial flagged CHECK in the manifest.
import { mkdirSync, existsSync, writeFileSync, createWriteStream } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import { Readable } from 'node:stream'

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..')
const DL = process.env.DOWNLOADED_PAGES_DIR || join(ROOT, 'downloaded_pages')
const IMG_DIR = join(DL, 'themealdb_images')
const API = 'https://www.themealdb.com/api/json/v1/1'
mkdirSync(IMG_DIR, { recursive: true })

const getJson = async (u) => { const r = await fetch(u, { headers: { 'user-agent': 'platyplus' } }); return r.ok ? r.json() : { meals: null } }
async function download(url, dest) {
  for (let a = 0; a < 3; a++) try {
    const r = await fetch(url); if (!r.ok) return r.status === 404 ? false : (a === 2 ? false : await new Promise((res) => setTimeout(() => res(false), 300)))
    await new Promise((res, rej) => { const w = createWriteStream(dest); Readable.fromWeb(r.body).pipe(w); w.on('finish', res); w.on('error', rej) }); return true
  } catch { if (a === 2) return false }
  return false
}

const CAT = (c) => { const x = (c || '').toLowerCase(); if (x === 'breakfast') return 'breakfast'; if (/dessert|side|starter/.test(x)) return 'snack'; return 'dinner' }
const clean = (s) => (s || '').replace(/<[^>]+>/g, ' ').replace(/&nbsp;/gi, ' ').replace(/\s+/g, ' ').trim()
function mapMeal(m) {
  const ingredients = []
  for (let i = 1; i <= 20; i++) { const ing = (m[`strIngredient${i}`] || '').trim(); const me = (m[`strMeasure${i}`] || '').trim(); if (ing) ingredients.push(`${me} ${ing}`.trim()) }
  const steps = clean(m.strInstructions).split(/\r?\n|(?<=\.)\s+(?=[A-Z])/).map((s) => s.trim()).filter((s) => s.length > 3)
  const tags = [m.strCategory, m.strArea, ...((m.strTags || '').split(',').map((t) => t.trim()).filter(Boolean))].filter(Boolean)
  const diet = []; if (/vegan/i.test(m.strCategory)) diet.push('vegan'); else if (/vegetarian/i.test(m.strCategory)) diet.push('vegetarian')
  return {
    id: 'tmdb-' + m.idMeal, title: m.strMeal, category: CAT(m.strCategory),
    minutes: 0, kcal: 0, protein: 0, carbs: 0, fat: 0,
    ingredients, steps, thumbnail: `/media/images/themealdb_images/${m.idMeal}.jpg`,
    tags, servings: 0, diet, source: 'themealdb',
  }
}

const seen = new Map()
for (const ch of 'abcdefghijklmnopqrstuvwxyz') {
  const { meals } = await getJson(`${API}/search.php?f=${ch}`)
  for (const m of meals || []) if (!seen.has(m.idMeal)) seen.set(m.idMeal, m)
  process.stdout.write(`  ${ch}:${seen.size} `)
}
console.log(`\nthemealdb: ${seen.size} meals`)

let got = 0
const recipes = []
for (const m of seen.values()) {
  const dest = join(IMG_DIR, m.idMeal + '.jpg')
  if (!existsSync(dest) && m.strMealThumb) { if (await download(m.strMealThumb, dest)) got++ } else if (existsSync(dest)) got++
  recipes.push(mapMeal(m))
}
writeFileSync(join(DL, 'themealdb_recipes.json'), JSON.stringify(recipes))
console.log(`images: ${got} -> ${IMG_DIR}; recipes -> downloaded_pages/themealdb_recipes.json`)
